src/libfetchers/zip-input-accessor.cc

```diff
@@ -0,0 +1,174 @@
+#include "input-accessor.hh"
+
+#include <zip.h>
+
+namespace nix {
+
+struct cmp_str
+{
+    bool operator ()(const char * a, const char * b) const
+    {
+        return std::strcmp(a, b) < 0;
+    }
+};
+
+struct ZipMember
+{
+    struct zip_file * p = nullptr;
+    ZipMember(struct zip_file * p) : p(p) { }
+    ~ZipMember() { if (p) zip_fclose(p); }
+    operator zip_file *() { return p; }
+};
+
+struct ZipInputAccessor : InputAccessor
+{
+    CanonPath zipPath;
+    struct zip * zipFile = nullptr;
+
+    typedef std::map<const char *, struct zip_stat, cmp_str> Members;
+    Members members;
+
+    ZipInputAccessor(const CanonPath & _zipPath)
+        : zipPath(_zipPath)
+    {
+        int error;
+        zipFile = zip_open(zipPath.c_str(), 0, &error);
+        if (!zipFile) {
+            char errorMsg[1024];
+            zip_error_to_str(errorMsg, sizeof errorMsg, error, errno);
+            throw Error("couldn't open '%s': %s", zipPath, errorMsg);
+        }
+
+        /* Read the index of the zip file and put it in a map.  This
+           is unfortunately necessary because libzip's lookup
+           functions are O(n) time. */
+        struct zip_stat sb;
+        zip_uint64_t nrEntries = zip_get_num_entries(zipFile, 0);
+        for (zip_uint64_t n = 0; n < nrEntries; ++n) {
+            if (zip_stat_index(zipFile, n, 0, &sb))
+                throw Error("couldn't stat archive member #%d in '%s': %s", n, zipPath, zip_strerror(zipFile));
+            auto slash = strchr(sb.name, '/');
+            if (!slash) continue;
+            members.emplace(slash, sb);
+        }
+    }
+
+    ~ZipInputAccessor()
+    {
+        if (zipFile) zip_close(zipFile);
+    }
+
+    std::string _readFile(const CanonPath & path)
+    {
+        auto i = members.find(((std::string) path.abs()).c_str());
+        if (i == members.end())
+            throw Error("file '%s' does not exist", path);
+
+        ZipMember member(zip_fopen_index(zipFile, i->second.index, 0));
+        if (!member)
+            throw Error("couldn't open archive member '%s' in '%s': %s",
+                path, zipPath, zip_strerror(zipFile));
+
+        std::string buf(i->second.size, 0);
+        if (zip_fread(member, buf.data(), i->second.size) != (zip_int64_t) i->second.size)
+            throw Error("couldn't read archive member '%s' in '%s'", path, zipPath);
+
+        return buf;
+    }
+
+    std::string readFile(const CanonPath & path) override
+    {
+        if (lstat(path).type != tRegular)
+            throw Error("file '%s' is not a regular file");
```

Comment:
Very strange, the error is with flake.lock:

```
error: file '/flake.lock' is not a regular file

       … while updating the flake input 'mach-nix'

       … while updating the lock file of flake 'git+file:///my/project/dir?ref=refs%2fheads%2fmy-current-branch'

```

Reply:
Oh, mach-nix has a symlink for the flake.lock. This explains the error. 

https://github.com/DavHau/mach-nix/blob/master/flake.lock

Reply:
I guess readFile has to resolve symlinks now. This in turns asks the question of 1) do we want to support that, 2) Should we prevent symlinks that go out of the flake boundary and 3) do we accept that lazy flakes (that live in a zip archive) will have a different behavior regarding symlinks that escape the flake than flakes that live plainly in the store.

Reply:
@edolstra This patch works for me :partying_face: : https://github.com/layus/nix/commit/44d5bd4789ded74eac6f717bd5c6686908f75ac6

I think it is safe because paths are rooted at the archive root. But is it always correct ? Could be.